lib/cdo/contact_rollups/v2/pardot.rb

```diff
@@ -132,50 +195,73 @@ def self.convert_to_prospect_fields(contact)
     prospects_payload_json_encoded = prospects_payload_json_encoded.gsub("+", "%2B")
 
-    "#{base_url}?prospects=#{prospects_payload_json_encoded}"
+    "#{api_endpoint}?prospects=#{prospects_payload_json_encoded}"
   end
 
   # Submits a request to Pardot to create/update a batch of prospects.
-  # This method may raise an exception.
   # @see http://developer.pardot.com/kb/api-version-4/prospects/#endpoints-for-batch-processing
   #
-  # @param base_url [String] a Pardot API endpoint
-  # @param prospects [Array<Hash>] array of prospect data
-  # @return [Array] array of errors containing indexes of rejected prospects and error messages
-  def self.submit_batch_request(base_url, prospects)
+  # @param api_endpoint [String] a Pardot API endpoint
+  # @param prospects [Array<Hash>] an array of prospect data
+  # @return [Array<Hash>] @see extract_batch_request_errors method
+  #
+  # @raise [Net::ReadTimeout] if doesn't get a response from Pardot
+  def self.submit_batch_request(api_endpoint, prospects)
     return [] unless prospects.present?
 
     # Send request to Pardot
-    url = build_batch_url base_url, prospects
-    time_start = Time.now
+    url = build_batch_url api_endpoint, prospects
+    time_start = Time.now.utc
     doc = post_with_auth_retry url
-    time_elapsed = Time.now - time_start
+    time_elapsed = Time.now.utc - time_start
 
     # Return indexes of rejected emails and their error messages
     errors = extract_batch_request_errors doc
 
-    log "Completed a batch request to #{base_url} in #{time_elapsed.round(2)} secs. "\
+    log "Completed a batch request to #{api_endpoint} in #{time_elapsed.round(2)} secs. "\
       "#{prospects.length} prospects submitted. #{errors.length} rejected."
 
     errors
   end
 
   # Extracts errors from a Pardot response.
-  # @param [Nokogiri::XML] doc Pardot XML response for a batch request
-  # @return [Array] array of hashes, each containing an index and an error message
+  # @param [Nokogiri::XML] doc a Pardot XML response for a batch request
+  # @return [Array<Hash>] an array of hashes, each containing an index and an error message
   def self.extract_batch_request_errors(doc)
     doc.xpath('/rsp/errors/*').map do |node|
       {prospect_index: node.attr("identifier").to_i, error_msg: node.text}
     end
   end
+
+  # Calculates what needs to change to transform an old data to a new data.
+  # @param [Hash] old_data
+  # @param [Hash] new_data
+  # @return [Hash]
+  def self.calculate_data_delta(old_data, new_data)
+    return new_data unless old_data.present?
+
+    # Set key-value pairs that exist only in the new data
+    delta = {}
+    new_data.each_pair do |key, val|
+      delta[key] = val unless old_data.key?(key) && old_data[key] == val
+    end
+
+    # Unset entries that exist only in the old data and not in the new data.
+    # Ignore entries with values are nil.
+    old_data.each_pair do |key, val|
+      delta[key] = nil unless new_data.key?(key) || val.nil?
```

Comment:
Can't wrap my head around this logic right now if anyone else can take a stab :)

Reply:
I'll add an example to the method comment to make it clearer.